sqlmesh/integrations/github/cicd/controller.py

```diff
@@ -0,0 +1,717 @@
+from sqlmesh.core.model import parse_model_name
+from sqlmesh.core.model.meta import IntervalUnit
+from sqlmesh.core.plan import Plan
+from sqlmesh.core.snapshot.definition import (
+    Intervals,
+    SnapshotChangeCategory,
+    _format_date_time,
+    merge_intervals,
+)
+from sqlmesh.core.user import User
+from sqlmesh.utils.date import make_inclusive
+from sqlmesh.utils.errors import CICDBotError, PlanError
+from sqlmesh.utils.pydantic import PydanticModel
+
+if t.TYPE_CHECKING:
+    from github import Github
+    from github.CheckRun import CheckRun
+    from github.Issue import Issue
+    from github.PullRequest import PullRequest
+    from github.PullRequestReview import PullRequestReview
+    from github.Repository import Repository
+
+    from sqlmesh.core.config import Config
+    from sqlmesh.core.snapshot import Snapshot
+
+logger = logging.getLogger(__name__)
+
+
+class PullRequestInfo(PydanticModel):
+    """Contains information related to a pull request that can be used to construct other objects/URLs"""
+
+    owner: str
+    repo: str
+    pr_number: int
+
+    @property
+    def full_repo_path(self) -> str:
+        return "/".join([self.owner, self.repo])
+
+    @classmethod
+    def create_from_pull_request_url(cls, pull_request_url: str) -> PullRequestInfo:
+        _, _, _, _, owner, repo, _, pr_number = pull_request_url.split("/")
+        return cls(
+            owner=owner,
+            repo=repo,
+            pr_number=pr_number,
+        )
+
+
+class GithubCheckStatus(str, Enum):
+    QUEUED = "queued"
+    IN_PROGRESS = "in_progress"
+    COMPLETED = "completed"
+
+    @property
+    def is_queued(self) -> bool:
+        return self == GithubCheckStatus.QUEUED
+
+    @property
+    def is_in_progress(self) -> bool:
+        return self == GithubCheckStatus.IN_PROGRESS
+
+    @property
+    def is_completed(self) -> bool:
+        return self == GithubCheckStatus.COMPLETED
+
+
+class GithubCheckConclusion(str, Enum):
+    SUCCESS = "success"
+    FAILURE = "failure"
+    NEUTRAL = "neutral"
+    CANCELLED = "cancelled"
+    TIMED_OUT = "timed_out"
+    ACTION_REQUIRED = "action_required"
+    SKIPPED = "skipped"
+
+    @property
+    def is_success(self) -> bool:
+        return self == GithubCheckConclusion.SUCCESS
+
+    @property
+    def is_failure(self) -> bool:
+        return self == GithubCheckConclusion.FAILURE
+
+    @property
+    def is_neutral(self) -> bool:
+        return self == GithubCheckConclusion.NEUTRAL
+
+    @property
+    def is_cancelled(self) -> bool:
+        return self == GithubCheckConclusion.CANCELLED
+
+    @property
+    def is_timed_out(self) -> bool:
+        return self == GithubCheckConclusion.TIMED_OUT
+
+    @property
+    def is_action_required(self) -> bool:
+        return self == GithubCheckConclusion.ACTION_REQUIRED
+
+    @property
+    def is_skipped(self) -> bool:
+        return self == GithubCheckConclusion.SKIPPED
+
+
+class AffectedEnvironmentModel(PydanticModel):
```

Comment:
why do you even need this class? why not enhance MissingInterval?

Reply:
https://github.com/TobikoData/sqlmesh/pull/810#discussion_r1186280280